src/material-experimental/theming/_theming.scss

```diff
@@ -0,0 +1,142 @@
+@use 'sass:string';
+@use '@angular/material' as mat;
+
+// Whether to throw an error when a required dep is not configured. If false, the dep will be
+// automatically configured instead.
+$_error-on-missing-dep: false;
+
+// Applies the theme for the given component configuration.
+@mixin _apply-theme($tokens, $component) {
+  $id: map.get($component, id);
+  $tokens: map.deep-merge($tokens, map.get($component, customizations));
+
+  // NOTE: for now we use a hardcoded if-chain, but in the future when first-class mixins are
+  // supported, the configuration data will contain a reference to its own theme mixin.
+  @if $id == 'mat.card' {
+    @include mat.card-theme($tokens);
+  } @else if $id == 'mat.checkbox' {
+    @include mat.checkbox-theme($tokens);
+  } @else {
+    @error 'Unrecognized component theme: #{id}';
+  }
+}
+
+// Gets the transitive dependency configurations for the given list of component configurations.
+@function _get-transitive-deps($components, $configured: ()) {
+  // Mark the given components as configured.
+  @each $component in $components {
+    $configured: map.set($configured, map.get($component, id), true);
+  }
+  $new-deps: ();
+
+  // Check each of the given components for new deps.
+  @each $component in $components {
+    @each $dep-getter in mat.private-normalize-args-list(map.get($component, deps)) {
+      $dep: meta.call($dep-getter);
+      $dep-id: map.get($dep, id);
+      @if not (map.has-key($configured, $dep-id)) {
+        @if $_error-on-missing-dep {
+          @error 'Missing theme: `#{map.get($component, id)}` depends on `#{$dep-id}`.' +
+            ' Please configure the theme for `#{$dep-id}` in your call to `mat.theme`';
+        } @else {
+          $configured: map.set($configured, $dep-id, true);
+          $new-deps: list.append($new-deps, $dep);
+        }
+      }
+    }
+  }
+
+  // Append on the new deps to this list of component configurations and return.
+  @if list.length($new-deps) > 0 {
+    $components: list.join($components, _get-transitive-deps($new-deps, $configured));
+  }
+  @return $components;
+}
+
+// Applies a theme using the given fallback tokens and component configurations.
+@mixin _theme($tokens, $components) {
+  $tokens: map.merge($tokens, (_mat-theme-type: 'tokens'));
+
+  // Save the original selector when back-compat is turned on. This lets us avoid nesting styles
+  // generated from the legacy theme in the default `body` selector.
+  @if map.has-key($tokens, back-compat) {
+    $tokens: map.deep-merge($tokens, (back-compat: (root-selector: &)));
+  }
+
+  // Call the theme mixin for each configured component.
+  @at-root #{& or body} {
+    @each $component in $components {
+      @include _apply-theme($tokens, $component);
+    }
+  }
+}
+
+// Takes the full list of tokens and a list of components to configure, and outputs all theme
+// tokens for the configured components.
+@mixin theme($tokens: mat.m2-tokens-from-theme(), $components) {
```

Comment:
Instead of having `$tokens` as a parameter here, do we want to make it just another node in the list?

Reply:
I know we did discuss this, but I felt a little hesitant because it behaves differently than the other functions. The other functions all request specific tokens to actually be emitted, this one would just be setting fallback values for things in case they are requested to be emitted.

So this `@include mat.theme((mat.m2-base-tokens($my-theme)))` would emit nothing because no components were requested to be emitted. Totally willing to brainstorm more on this, that was just my reasoning for leaving it separate in this PR